lib/utils.py

```diff
@@ -141,22 +148,23 @@ def is_hostname(string):
         """
 
         try:
-            return bool(re.match("^(([a-zA-Z]|[a-zA-Z][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z]|[A-Za-z][A-Za-z0-9\-]*[A-Za-z0-9])$", string))
+            return bool(re.match("^(([a-zA-Z]|[a-zA-Z][a-zA-Z0-9-]*[a-zA-Z0-9])\\.)*([A-Za-z]|[A-Za-z][A-Za-z0-9\-]*[A-Za-z0-9])$", string))
         except TypeError:
             return False
 
     @staticmethod
     def get_local_ipv4_address():
         """
-        Get's local ipv4 address
-        TODO: What if more than one interface present ?
+        Get local ipv4 address of the interface with the default gateway.
+        Return '127.0.0.1' if no suitable interface is found
+        TODO: What if more than one interface present? What if 127.0.0.1 is not active?
 
         :return: IPv4 address as a string
         :rtype: string
         """
         try:
             s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-            s.connect(('10.255.255.255', 1))
+            s.connect(('8.8.8.8', 1))
```

Comment:
Was ist der Benefit/der inhaltliche Unterschied der Änderung? 
Ich habe Bedenken, dass unbedarfte Leser dabei denken, dass shng mit Google kommuniziert.

Reply:
Auch aus der alten lib.network von Foxi.

Kann ich sonst nix zu sagen. 8.8.8.8 ist der Google-Resolver, das stimmt. Gefällt mir auch nicht.

Reply:
Ich hab nochmal im Netz geschaut. Die meisten Beispiele verwenden 8.8.8.8, allerdings gibt es einige Beispiele, die auf 10.255.255.255 setzen mit dem Hinweis, die Adresse müsse nicht erreichbar sein. 

Bei mir geht es auch, obwohl kein 10er Netzwerk erreichbar ist (ist aus dem Bereich der 'private IP', wird also nicht geroutet). Da fände ich die originale Lösung mit 10.255.255.255 besser.

@Foxi352 kannst du ggf. erläutern, wieso du die Google-Adresse genommen hast? Ging damit bei dir etwas nicht (Windows?)

Reply:
Da es schon relativ lange her ist kann ich mich auch nicht mehr genau erinnern. Ich denke aber es war folgendes:

Ich denke, dass ich Probleme hatte dass ich Interfaces in einem 10er Netz hatte die somit als "default" mit "gateway" erkannt wurden weil der Connect funktionierte. Es war aber das falsche Interface.

Das Problem ist vor allem wenn du Hypervisor, Container engines, VPN Clients e.t.c. nutzt. (Oder wir ich alles zusammen). Ich hab auf meinem Mac hier auf der Arbeit z.B. 19 Netzwerk interfaces.

Wenn man ein Connect auf etwas macht was wirklich im Internet ist dann kann man relativ sicher sein, dass die auch das Default Interface ist. Ich wusste nicht was ich nutzen sollte was auch relativ sicher permanent da ist und nicht ändert. Wahrscheinlich ware die Google Resolver das beste was mir eingefallen war :-) Man könnte ja auch die SHNG Web IP nehmen. Aber dann denken die unbedarften Leser dass wir ein phone-home machen.

Wenn jemand eine bessere Lösung für das Problem hätte gefiele mir auch besser.

Reply:
Ich vermute, es ist egal, welche "externe" IP man nimmt, irgendwem gefällt es immer nicht, egal ob Google 8.8.8.8, Cloudflare 1.1.1.1, shng WebIP oder sonst eine. Man wird sich dann auch realistisch nicht von "irgendeinem" Vorwurf freimachen können.

Andererseits funktioniert die Prüfung nur zuverlässig, wenn tatsächlich eine externe IP verwendet wird. Ich möchte jetzt auch keine "offizielle" IP, z.B. von einem DNS rootserver nutzen.

@Foxi352: das Problem, dass bei dir eine gültige IP ausgewählt wurde, aber nicht die, die du gerade brauchst, lässt sich nicht automatisiert lösen. Wie auch. Ob du gerade die externe oder eine der Container-/VM-IPs brauchst, kann keine Automatik wissen.

Wenn bei dir die Lösung mit 10.255.255.255 "funktioniert" (gültige IP zurückgibt), auch wenn es - für dich - nicht die gewünschte ist, würde ich die Möglichkeit bevorzugen.
Wenn bei dir trotz gültiger nicht-127.0.0.1-IP trotzdem nur 127.0.0.1 rauskommt, wäre das ein no-go.

Reply:
Ich möchte hier nochmal etwas ergänzen.

1. Ich habe jetzt insgesamt fast 2 Stunden das Netz durchforstet - es gibt keine Routine, die plattformübergreifend, unabhängig von aktivem Internet und unabhängig von lokal konfiguriertem DNS die lokalen IP-Adresse(n) ermittelt.

2. Die lib.utils.get_local_ipv4_address() wird von modules/admin und modules/websocket, weiters plugins/backend/BackendSysteminfo und plugins/nuki direkt verwendet. In modules/mqtt ist eine quasi-1:1-Kopie (mit Google 8.8.8.8) hart kodiert (kein Verweis auf lib.utils).

3. In den Fällen von 2) wird die "lokale" IPv4-Adresse verwendet, um "0.0.0.0" als Parameter zu ersetzen (Ausnahme ist backend/Systeminfo). Das halte ich inhaltlich für falsch. Wenn der default 0.0.0.0 ist (oder manuell gesetzt wurde), dann sollte die Spezifikation auch eingehalten und das Socket an "alle" lokalen Netzwerkinterfaces gekoppelt werden.
Wenn das für einzelne Anwendungen zu gefährlich ist, dann muss der Default eben anders lauten, z.B. 127.0.0.1.
Aber als default die 0.0.0.0 anzugeben und dann an eine einzelne, willkürlich, weil nicht steuerbar, ausgewählte IP zu binden, darf nicht vorkommen.

Insofern würde ich vorschlagen:

a) Die Version mit 10.255.255.255 nehmen, es sei denn, @foxi352 kann auf seinem System Fehlverhalten nachweisen (siehe oben). Ggf. könnte man die IP auch - wie @bmxp vorgeschlagen hat - als Parameter in der smarthome.yaml vorbelegen bzw. ändern für die, die damit ein Problem haben. Ich würde auf die 8.8.8.8 lieber ganz verzichten.

b) Das Verhalten der Module http, admin und mqtt sowie Plugin nuki entsprechend anpassen. Ob man dazu lieber als (sicheren) Default localhost nimmt, oder dann alle Interfaces nutzt, ist eine Designentscheidung. Das muss dann auch in der Doku klar gestellt werden.

c) Wenn entschieden ist, wie das umgesetzt werden soll, kann ich das gern alles entsprechend anpassen und hier im PR mit einbauen.

Meinungen?

Reply:
Wenn das für alle passt, dann nehmt keine Rücksicht auf mich. Ich hatte das damals vermutlich geändert weil ich irgendwo ein Problem hatte, ansonsten hätte ich das wohl nicht angefasst. Getestet und entwickelt hatte ich damals auf Linux, MacOS und FreeBSD. Wo ich welche Probleme hatte weiss ich nicht mehr, es ist zu viel Zeit zwischen der Entwicklung und der wirklichen Nutzung der Lib vergangen.

Wenn's aber für alle Linux Nutzer, und somit wahrscheinlich 99% der SHNG User, passt dann würde ich das wieder zurück ändern zu 10.255.255.255. Sollte ich je wieder das Problem entdecken werde ich es im Detail untersuchen und dann können wir immer noch eine andere Lösung suchen. Mein production System läuft inzwischen in einem Debian LXC Container unter Proxmox. Die Änderung dürfte mir also keine Schwierigkeiten machen.

Bei Punkt 3 bin ich ganz bei dir was die 0.0.0.0 betrifft 0.0.0.0 ist für mich auch ganz klar "bind to all".

Ich würde es gut finden, wie schon von einigen vorgeschlagen, einen optionalen Interface Parameter in der shng.yaml einzuführen. Ist dieser gesetzt kann die get_local_ipv4 einfach die vom Nutzer festgelegte Adresse zurückgeben. Ist der Parameter nicht gesetzt läuft die Methode normal durch und gibt halt das zurück was sie meint. Sollte dann jemand mit der 10.255.255.255 ein Problem haben kann er halt den Parameter nutzen um das IF selbst anzugeben und gut ist ..